pkg/services/auth/jwt/user_auth.go

```diff
@@ -0,0 +1,238 @@
+		return err
+	}
+
+	return nil
+}
+
+type UserAuthService struct {
+	Cfg *setting.Cfg
+
+	keySet keySet
+	log    log.Logger
+
+	additionalClaims map[string]interface{}
+	expectedClaims   jwt.Expected
+
+	remoteCache         *remotecache.RemoteCache
+	verificationService VerificationService
+}
+
+func (s *UserAuthService) Verify(ctx context.Context, token string) (models.JWTClaims, error) {
+	return s.verificationService.Verify(ctx, s.keySet, s.expectedClaims, s.additionalClaims, token)
+}
+
+func (s *UserAuthService) initClaimExpectations() error {
+	if err := json.Unmarshal([]byte(s.Cfg.JWTAuthExpectClaims), &s.additionalClaims); err != nil {
+		return err
+	}
+
+	for key, value := range s.additionalClaims {
+		switch key {
+		case "iss":
+			if stringValue, ok := value.(string); ok {
+				s.expectedClaims.Issuer = stringValue
+			} else {
+				return fmt.Errorf("%q expectation has invalid type %T, string expected", key, value)
+			}
+			delete(s.additionalClaims, key)
+		case "sub":
+			if stringValue, ok := value.(string); ok {
+				s.expectedClaims.Subject = stringValue
+			} else {
+				return fmt.Errorf("%q expectation has invalid type %T, string expected", key, value)
+			}
+			delete(s.additionalClaims, key)
+		case "aud":
+			switch value := value.(type) {
+			case []interface{}:
+				for _, val := range value {
+					if v, ok := val.(string); ok {
+						s.expectedClaims.Audience = append(s.expectedClaims.Audience, v)
+					} else {
+						return fmt.Errorf("%q expectation contains value with invalid type %T, string expected", key, val)
+					}
+				}
+			case string:
+				s.expectedClaims.Audience = []string{value}
+			default:
+				return fmt.Errorf("%q expectation has invalid type %T, array or string expected", key, value)
+			}
+			delete(s.additionalClaims, key)
+		}
+	}
+
+	return nil
+}
+
+func (s *UserAuthService) checkKeySetConfiguration() error {
+	var count int
+	if s.Cfg.JWTAuthKeyFile != "" {
+		count++
+	}
+	if s.Cfg.JWTAuthJWKSetFile != "" {
+		count++
+	}
+	if s.Cfg.JWTAuthJWKSetURL != "" {
+		count++
+	}
+
+	if count > 1 {
+		return ErrKeySetConfigurationAmbiguous
+	}
+
+	if count == 0 {
+		return ErrKeySetIsNotConfigured
+	}
+
+	return nil
+}
+
+func (s *UserAuthService) initKeySet() error {
+	if err := s.checkKeySetConfiguration(); err != nil {
+		return err
+	}
+
+	if keyFilePath := s.Cfg.JWTAuthKeyFile; keyFilePath != "" {
+		// nolint:gosec
+		// We can ignore the gosec G304 warning on this one because `fileName` comes from grafana configuration file
+		file, err := os.Open(keyFilePath)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := file.Close(); err != nil {
+				s.log.Warn("Failed to close file", "path", keyFilePath, "err", err)
+			}
+		}()
+
+		data, err := io.ReadAll(file)
+		if err != nil {
+			return err
+		}
+		block, _ := pem.Decode(data)
+		if block == nil {
+			return ErrFailedToParsePemFile
+		}
+
+		var key interface{}
+		switch block.Type {
+		case "PUBLIC KEY":
```

Comment:
Nice! Are we planning on sharing the public key with the plugin so it can verify that the JWT is issued by grafana?

Reply:
I think this service is the existing [JWT auth](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/jwt/) service and has nothing to do with generating JWT's for plugins.

Reply:
Ah - mybad 🙈